### PR TITLE
feat(notification): support send notification when schedule job failed

### PR DESCRIPTION
**Pull request by LuckyPickleZZ**
#### What type of PR is this?
type-feature

#### What this PR does / why we need it:
1. Support sending notifications when schedule job scheduling execution fails
2. Change the matching rule of notification policy from string matching to map matching
3. Fixed a bug where if a notification within a batch fails to be sent, the entire transaction will be rolled back, resulting in multiple message resents

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/quartz/OdcJobListener.java [hunk 1]
```diff
@@ -127,6 +137,26 @@ public void jobExecutionVetoed(JobExecutionContext context) {
 
     @Override
     public void jobWasExecuted(JobExecutionContext context, JobExecutionException jobException) {
+        if (jobException != null) {
```
Comment (LuckyPickleZZ) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/quartz/OdcJobListener.java:140]: if notificationEnabled

**Review by MarkPotato777 — COMMENTED** [review]
I have some questions that we can discuss:
- Would it be better if we do not enqueue events while there is no notification policy configured in the current organization? I doubt if all events are pushed into the event queue, the conversion from event to notification will be delayed and so does the notification sending.
- Do we need a pressure test to make sure of the stability? 
- Do we need to refine the configuration values of NotificationProperties? or add some other configurations?

**Comment by LuckyPickleZZ** [comment]
> I have some questions that we can discuss:
> 
> * Would it be better if we do not enqueue events while there is no notification policy configured in the current organization? I doubt if all events are pushed into the event queue, the conversion from event to notification will be delayed and so does the notification sending.
> * Do we need a pressure test to make sure of the stability?
> * Do we need to refine the configuration values of NotificationProperties? or add some other configurations?

As a temporary solution, I believe that all of the above issues can be solved by reducing the scheduling period of periodic tasks.
If we want to completely solve it, we may be able to schedule events and notifications as asynchronous tasks in periodic tasks.

**Comment by MarkPotato777** [comment]
> > I have some questions that we can discuss:
> > 
> > * Would it be better if we do not enqueue events while there is no notification policy configured in the current organization? I doubt if all events are pushed into the event queue, the conversion from event to notification will be delayed and so does the notification sending.
> > * Do we need a pressure test to make sure of the stability?
> > * Do we need to refine the configuration values of NotificationProperties? or add some other configurations?
> 
> As a temporary solution, I believe that all of the above issues can be solved by reducing the scheduling period of periodic tasks. If we want to completely solve it, we may be able to schedule events and notifications as asynchronous tasks in periodic tasks.

okay, you could conduct a quick test to confirm the scheduling and batch-size configurations~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/notification/NotificationSchedules.java [hunk 2]
```diff
@@ -40,6 +41,7 @@ public class NotificationSchedules {
     @Autowired
     private NotificationProperties notificationProperties;
 
+    @Async
```
[the remark this answers is not in the document]
Comment (LuckyPickleZZ) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/notification/NotificationSchedules.java]: done

**Comment by LuckyPickleZZ** [comment]
mocked 100W records for `notification_event` table, and fetch 1W records once.

image

**Review by MarkPotato777 — APPROVED** [review]
LGTM

**Pull request merged**